### ISSUE TITLE
Homespan Device unresponsive in Homekit.

### ISSUE THREAD

**Issue by prolixmusic** [HomeSpan/HomeSpan#791]
I built an ESP32 based garage door operator. This afternoon (after 10 days of running flawlessly) it lost homekit connectivity. My Home reports 'no response' for the garage door, whereas everything else (thermostat, door locks, etc.) are fine. I pinged my garage door opener, and it responded, but with wildly varying round-trip times. That's a little concerning given that it's less than 10ft from my access point. Never the less, it is on the network. 

In my sketch, everything happens in void loop() in DEV_Garagedoor.h, which I presume is called by homespan.  Anyway, each time that is called, I check timeouts for the light and the door, check for inputs from hardwired buttons for the door and light, and run a state for managing the open/closes cycles.

Interestingly, despite, my iPhone reporting 'no response' from the garage door, the controller still responds properly to the hardwired inputs - opens and closes to door, turns the lights on and off, etc. which means that the sketch is still running.

So, it's on the network, and the sketch is running... does anyone have any idea what might be causing this? Is there a homespan method that I would allow me to reset the controller in the event it loses connection with My Home?

Any suggestion would be appreciated.

**Comment by HomeSpan** [comment]
These types of connectivity issues are both frustrating as well as difficult to solve, especially if everything is working fine and then just stops.

I assume you tried simply rebooting the ESP32 - if not, that's the first step.  If you don't have weblog enabled, add it to your sketch so you can readily check that the ESP32 is still on the network.  The weblog will also provide info on how often it loses and re-connects to WiFi.

When accessing the weblog, try using both the hostname as well as the IP address.  If the IP address is responsive but the hostname is not, this suggests an MDNS issue with how your router may be handling those requests.

Are you using a Mesh network or just a single AP?  Mesh networks can often dynamically reconfigure themselves, sometimes to the detriment of a specific device.  Also, WiFi can be very sensitive to the local environment - a neighbor may have changed their own WiFi, which could degrade the performance of your system is channels overlap.

If your router has an automatic channel optimizer, try running that.  If that doesn't work, try rebooting the router as well.

One final diagnostic is to check if the issue is related to HomeSpan.  You can check this by uploading one of the ESP32  WebServer examples that use MDNS.  If that sketch is showing laggy response, then the problem is likely a WiFi access issue.  As a temporary check, you can try moving the board to another location and see if things work any better.

**Comment by prolixmusic** [comment]
I have not tried rebooting yet.  Of course, that will work.  I have a single AP for WiFi that is on the same floor as the garage and literally 10ft from the esp32. The esp32 is online but is not broadcasting mDNS any more.  Is there a way to monitor the status of the mDNS service  and reboot the ESP if the service stops running?  Sent from my iPhoneOn Feb 29, 2024, at 10:46 PM, HomeSpan ***@***.***> wrote:﻿
These types of connectivity issues are both frustrating as well as difficult to solve, especially if everything is working fine and then just stops.
I assume you tried simply rebooting the ESP32 - if not, that's the first step.  If you don't have weblog enabled, add it to your sketch so you can readily check that the ESP32 is still on the network.  The weblog will also provide info on how often it loses and re-connects to WiFi.
When accessing the weblog, try using both the hostname as well as the IP address.  If the IP address is responsive but the hostname is not, this suggests an MDNS issue with how your router may be handling those requests.
Are you using a Mesh network or just a single AP?  Mesh networks can often dynamically reconfigure themselves, sometimes to the detriment of a specific device.  Also, WiFi can be very sensitive to the local environment - a neighbor may have changed their own WiFi, which could degrade the performance of your system is channels overlap.
If your router has an automatic channel optimizer, try running that.  If that doesn't work, try rebooting the router as well.
One final diagnostic is to check if the issue is related to HomeSpan.  You can check this by uploading one of the ESP32  WebServer examples that use MDNS.  If that sketch is showing laggy response, then the problem is likely a WiFi access issue.  As a temporary check, you can try moving the board to another location and see if things work any better.

**Comment by frankonski** [comment]
I have seen the “No response” happen when the Home app receives incorrect json status data from HomeSpan (that is when I was trying to add extra functionality..).

Review all your “setVal()” methods and ensure the values passed are valid and accepted ones for the characteristic.

If you have weblogs enabled you could also see something there.

**Comment by prolixmusic** [comment]
So I did some debugging with Wireshark. There is definitely something awry with homespans MDNS. 

When you reset homespand, it posts half a dozen MDNS messages that look similar to those that are routinely posted by my Lyric... 

622	29.785617	192.168.0.156	224.0.0.251	MDNS	219	Standard query 0x0000 ANY Spaceship Dock._hap._tcp.local, "QM" question ANY HomeSpan-07A74546A0D8.local, "QM" question ANY HomeSpan-07A74546A0D8.local, "QM" question SRV 0 0 80 HomeSpan-07A74546A0D8.local A 192.168.0.156

After that, the only time it posts any MDNS messages is if I run the Discovery App on my phone which queries bonjour devices... The first time you run the app after a reboot you get a similar message to the one originally posted.  On subsequent attempts at running the discovery app, it posts an MDNS message that is pretty much empty.

789	35.319705	192.168.0.156	224.0.0.251	MDNS	97	Standard query response 0x0000 A, cache flush 192.168.0.156

Then, after a day or 2 it disappears altogether.

This seems to be an issue with MDNS...

Thoughts??

**Comment by HomeSpan** [comment]
Various users have reported similar issues with MDNS.  It is unknown whether this is a problem with the ESP32 MDNS library itself, the way that some routers handle MDNS traffic, or a combination of both.  This issue apparently extends beyond HomeSpan to commercial devices as well, and it seems very idiosyncratic and specific to the WiFi router used, especially if it's a mesh network.

I used a Linksys Velop mesh for a few years and everything worked fine.  Then after a variety of firmware updates to the Linksys router, things started to decay - both for my HomeSpan devices as well as any commercial HomeKit devices that relied on WiFi.  After a while I tossed the Linksys Velop and replaced with a Netgear Orbi mesh and everything worked immediately (and has been for about 2 years).

As you can see from the Issues list, I've worked with other users reporting similar problems on different ways of debugging, restarting MDNS, etc., but unfortunately we were not able to to find a general solution.

One thing you may want to test is whether a non-HomeSpan sketch has the same issues.  The Arduino IDE contains a number of ESP32 examples, including a few for ESPmDNS you can try as a baseline.

What WiFi router system are you using?  Also, have you tried rebooting the router?

**Comment by prolixmusic** [comment]
I have a TPLink router and a ubiquiti access point. Of course the router is responsible for DNS within the lan, but mDNS is supposed to independent of that.  I have a ton of devices on my network that use bonjour (mDNS) and, can see all of them responding… the other HAP device on my network is a LYRIC thermostat -   It has been running flawlessly for 5+ years.  Its mDNS data packet is similar in format to Homespans original packet.  It feels like something is overwriting/ corrupting the mDNS data in RAM  — ie a firmware problem… I just can’t see why my router or AP would selectively strip the data from homespan… particularly when other HAP devices are unaffectedSent from my iPhoneOn Mar 7, 2024, at 11:22 PM, HomeSpan ***@***.***> wrote:﻿
Various users have reported similar issues with MDNS.  It is unknown whether this is a problem with the ESP32 MDNS library itself, the way that some routers handle MDNS traffic, or a combination of both.  This issue apparently extends beyond HomeSpan to commercial devices as well, and it seems very idiosyncratic and specific to the WiFi router used, especially if it's a mesh network.
I used a Linksys Velop mesh for a few years and everything worked fine.  Then after a variety of firmware updates to the Linksys router, things started to decay - both for my HomeSpan devices as well as any commercial HomeKit devices that relied on WiFi.  After a while I tossed the Linksys Velop and replaced with a Netgear Orbi mesh and everything worked immediately (and has been for about 2 years).
As you can see from the Issues list, I've worked with other users reporting similar problems on different ways of debugging, restarting MDNS, etc., but unfortunately we were not able to to find a general solution.
One thing you may want to test is whether a non-HomeSpan sketch has the same issues.  The Arduino IDE contains a number of ESP32 examples, including a few for ESPmDNS you can try as a baseline.
What WiFi router system are you using?  Also, have you tried rebooting the router?

**Comment by HomeSpan** [comment]
MDNS messages can be provided in different ways, and timing can be different -  there does not seem to be a standard and sometimes certain devices fail and others do not.  Have you tried testing with the Espressif ESPmDNS example?

Also, see https://community.tp-link.com/en/business/forum/topic/606830 for reports of similar problems with commercial devices.

**Comment by prolixmusic** [comment]
So I have a question/observation... I looked at HomeSpan.cpp and see the following statements pertaining to MDNS:
 
  MDNS.begin(hostName);                         // set server host name (.local implied)
  MDNS.setInstanceName(displayName);            // set server display name
  MDNS.addService("_hap","_tcp",tcpPortNum);    // advertise HAP service on specified port

When I looked at the ESP examples, they only use MDNS.begin and MDNS..addservice.. They don't use MDNS,setInstanceName

Just wondering if this has anything to do with the issues I am seeing??

**Comment by HomeSpan** [comment]
The instanceName is used to set a user-friendly display name in the MDNS record.  If you poll all your MDNS devices you'll see that they often have a user-friendly name that is different from the hostname.  

If you have not yet rebooted your router, I highly recommend trying this before proceeding much further, given that you indicated the ESP32 had wildly varying response times.   Also, check if your router has any setting related to MDNS, or any type of filtering, and disable them all.  These are known to cause MDNS fails with certain devices.

**Comment by prolixmusic** [comment]
I will take another look at the router…although it doesn’t seem to be blocking MDNS for anything else on the network.  I also looked at other MDNS libraries.  After the router I try another library. Sent from my iPhoneOn Mar 8, 2024, at 10:00 PM, HomeSpan ***@***.***> wrote:﻿
The instanceName is used to set a user-friendly display name in the MDNS record.  If you poll all your MDNS devices you'll see that they often have a user-friendly name that is different from the hostname.
If you have not yet rebooted your router, I highly recommend trying this before proceeding much further, given that you indicated the ESP32 had wildly varying response times.   Also, check if your router has any setting related to MDNS, or any type of filtering, and disable them all.  These are known to cause MDNS fails with certain devices.

**Comment by HomeSpan** [comment]
If you Google Ubiquiti and MDNS you'll find a lot of message board reporting similar issues as you have described, along with various potential fixes.

**Comment by prolixmusic** [comment]
I am pretty sure my AP is not an issue.  I have many bonjour devices throughout my home and they are all accessible and all work on my network via the ubiquiti AP. Also when I scan my network, again via WiFi, I get a list of devices with their capabilities… homespan only shows up as pingable, where as many others show up as bonjour capable. I think there has to be something awry with the mDNS on homespan. After its initial broadcast it then only responds with empty messages.  All the details are missing.  Sent from my iPhoneOn Mar 8, 2024, at 10:36 PM, HomeSpan ***@***.***> wrote:﻿
If you Google Ubiquiti and MDNS you'll find a lot of message board reporting similar issues as you have described, along with various potential fixes.

**Comment by HomeSpan** [comment]
Okay.  Please let me know how the debugging goes and if you find anything.  I would love to be able to resolve this but it's very hard for me to debug since I can't reproduce the error on my end.  And I'm trying to avoid writing my own MDNS handler 😄.

**Comment by prolixmusic** [comment]
This is for sure a little odd.  11am this morning it cold no longer be found.  I reset it and have had wireshark watching the MDNS traffic all day.  So far, with the exception of one malformed message it seems to be behaving.  When it’s running properly, my lan diagnostics app shows that it responds to http, bonjour and ping.  When it starts misbehaving, http and bonjour are no longer available, despite the fact that the sketch continues to respond properly to hardwired inputs. Sent from my iPhoneOn Mar 9, 2024, at 10:02 AM, HomeSpan ***@***.***> wrote:﻿
Okay.  Please let me know how the debugging goes and if you find anything.  I would love to be able to resolve this but it's very hard for me to debug since I can't reproduce the error on my end.  And I'm trying to avoid writing my own MDNS handler 😄.

**Comment by prolixmusic** [comment]
According to Espressifs mDNS documention, by default mDNS is confirgured as a low priority task. Espressif also has recommendations for optimizing application performance. One statement jumper out at me... 

"It is also necessary to ensure that tasks yield CPU (by calling [vTaskDelay()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos_idf.html#_CPPv410vTaskDelayK10TickType_t), sleep(), or by blocking on semaphores, queues, task notifications, etc) in order to not starve lower-priority tasks and cause problems for the overall system

So here's a question, are we staring lower priority tasks like mDNS?

BTW its still running today, although, I have noticed that that my homespan device receives queries from the My home hub at a much lower cadence than other devices -- Is this a function on the TTL value in the response record?

Thoughts?

**Comment by HomeSpan** [comment]
The Arduino-ESP32 library already has the `vTaskDelay()` built into main `loop()` function so should be okay.  Can you post a copy of your full sketch?

**Comment by prolixmusic** [comment]
[GarageDoor_copy_20240211193856.zip](https://github.com/HomeSpan/HomeSpan/files/14551358/GarageDoor_copy_20240211193856.zip)

Here is my garage door implementation... I built a custom board for  a Hiletgo ESP-WROOM-32 development board... I use normally closed door contacts that push the input high, that way is a contact opens, or a wire is broken, or a wire is shorted to ground, it fails safe as opposed to overriding the the switch. 

My loop should iterate quickly as there are no delay calls. I have a filter routine to debounce switches and reject couple noise that use an event counter as opposed to a delay. 

I use the current->getVal() and current->setVal() as the state both for Homespan and my state machine.

I use target->setVal() to set a new target value in Homekit when the door button on the wall is pressed. This simplifies the state machine and makes sure that homekit reflects the current targetted state. This will only get called once when the hardwired door button is pressed, so there should be negligible network traffic as a result. 

Anyway this was my first foray into sketching in the Arduino Environment, and first project time ESP32. I spend most of my life building networked mixed signal embedded systems with PIC's, AVR's, etc..

**Comment by HomeSpan** [comment]
Above you mention that when you lose the MDNS, you also note that the device no longer responds to http.  How are you testing this - by using the IP address or by using the hostname?   If you enable the web log in your sketch you should be able to access it and get log output even if MDNS fails as long as you use the IP address.  If that *does not* work, then the issue is more than MDNS.

**Comment by prolixmusic** [comment]
I have been testing use netAnalyzer on my iPhone to examine basic capabilities of devices on my local subnet. It scans IP addresses on the subnet and reports on whether an IP is pingable, has website and bonjour capability, amongst other things. . When the ESP stops responding, the app reports that the ESP's IP is ONLY pingable.

So this might be comical, but as I said before, my sketch doesn't use delays, so it execution of the loop should be really fast... My assumption was that a delay in Arduino would be blocking and therefore detrimental to the RTOS, but in another forum someone noted that the Arduino implementation of delay for ESP actually uses vTaskDelay() under the hood and is therefore non-blocking. Now I am wondering if my lack of delays may actually be starving the underlying stack tasks of processor time. 

BTW, thanks for supporting my troubleshooting... This has been really useful!

**Comment by prolixmusic** [comment]
BTW here are some of the many mDNS responses from the ESP32. Does the fact that they do not all contain the same data point to anything?


![Screenshot 2024-03-10 193908](https://github.com/HomeSpan/HomeSpan/assets/111787109/0569ca09-90f6-4d19-913a-53319afc70a4)

**Comment by HomeSpan** [comment]
You should never need to add a delay in a HomeSpan sketch.  It uses the Arduino-ESP32 library that includes the required delay.   Likewise, you should never run a task that blocks the loop from executing.  For example, if you are running a motor for 30 seconds, that needs to occur without blocking.

One other thing to check are your calls to `setVal()`.  These should be very sporadic and only occur when there is an actual state change that needs to be reported.  If the garage door is opening, you should set currentState to OPENING only once.  Flooding the network with continuous `setVal()` updates can certainly cause connectivity issues.

You should definitely enable the HomeSpan Web Log (simply add `homeSpan.enableWebLog()` to your sketch) and check if you can reach it from the browser with its IP address while the device appears to be working, and then again when your diagnostics show there is a problem.

If you can't reach the device via the weblog using its IP address then the issue is 99% NOT related to MDNS.  The weblog doesn't have anything to do with MDNS, and if you use the IP address to connect, your router will not need MDNS either. 
Note if you can reach the weblog that does not mean it's definitely an MDNS problem, but it is more likely. 

If possible, try to run your sketch while the device is connected to your computer - you may be able to use the Serial Monitor to get a better idea of what's going on.

**Comment by HomeSpan** [comment]
Re the MDNS logs, these look about the same as those on my own devices.  MDNS is quite complex and I've never really digested all the message types and how/when they should be broadcast.

**Comment by prolixmusic** [comment]
Yes there are no delays... Everything happens in a state machine. I use target->timeVal() for door and light timing. As far as I am aware that should not be making network calls to homekit.

Then I use target->getVal() to see what the current target is. 

The only time I use target->setVal() is when the physical door button is pressed --- on a weekday, this is twice a day!

Then I use current->setval() to set the current state. This is the same state that the state machine in my loop uses. I read the value with current->getVal() each time the loop runs.

I use current->setVal only when a state changes... So during a normal opening, the state starts out as closed, changes once to opening, then once the door is open changes to opened.

There are no blocking statement in the loop... no while statements!

Anyway, it should be really fast (it is very responsive), none blocking, and have low network overhead. 

Having found out that the Arduino ESP implementation of delay() actually uses vTaskDelay() under the hood, which is none blocking and should yield to other tasks, I wonder if the fact that I don't have any delays is starving the stack..

**Comment by HomeSpan** [comment]
You do not need to add any delays.  I have no delays in any of my sketches.  No one has any delays in any of their sketches.  Using delay in your sketch should be avoided at all costs except under limited circumstances where it is okay to block the operation of the main HomeSpan polling function.

**Comment by prolixmusic** [comment]
Yes, BUT delay in ESP is apparently  **non-blocking** and therefore yields to the OS... I get it, it sounds counterintuitive, but I see in the mDNS code that mDNS has a very low priority by default, and therefore may be less responsive than it ought to be...

That being the case adding a delay() (which under the hood is a vTaskDelay())may actually help with the responsiveness of the stack.

**Comment by prolixmusic** [comment]
BTW on mDNS i found the Wikipedia page to be quite helpful for the basics ...

So here you can see my Homekit Hub asking QUESTIONING (QM) Spaceship Dock, and Spaceship Dock responding with a partial message. You can tell by the 0x**** sequence number that Spaceship Dock is responding to a specific request versus something that is global
![Screenshot 2024-03-10 201250](https://github.com/HomeSpan/HomeSpan/assets/111787109/546e6b6b-63c7-4054-819b-a13ae9e79cf0)

Interestingly, here it respond to the same question with even more data. Why is one concatenated??
![Screenshot 2024-03-10 201506](https://github.com/HomeSpan/HomeSpan/assets/111787109/9d2727f4-2371-4556-92d8-76e9d2da8767)

I have times when the response to the same message is completely empty too... This makes no sense to me.
![Screenshot 2024-03-10 202843](https://github.com/HomeSpan/HomeSpan/assets/111787109/353415d4-ebb0-44b5-9e77-9d6a284adf40)


I am pretty sure that responses with sequence number 0x0000 are responses to a more generalized question for all devices to post data. That being the case, I think those packets should always contain the PTR, SRV and TXT fields.

**Comment by prolixmusic** [comment]
Interesting cadence to the ping responses…

**Comment by prolixmusic** [comment]
Tried posting this via email, the other day, but the image didn't post... Anyway, these are ping results for homespan... Anyway the cadence is interesting and not what I'd expect.
![Ping Statistics](https://github.com/HomeSpan/HomeSpan/assets/111787109/458bf8df-ae04-4ed2-bfbe-6c11e9360a88)
These are the results for my homeSpan sketch with no delays.  I then tried adding an additional vTaskDelay(5) before the homeSpan.poll() routine and did see a decrease in roundtrip time of around 20%, but the same sawtooth cadence coupled with trend for roundrtip times to increase overtime. I was hoping that I might see something more consistent like sub 10 ms round trip times, but unfortunately that was the case.

**Comment by Nezaemmy** [comment]
Hi, I had this issue in the past, When the WiFi RSSI went in the range between -90 and -94 the WiFi module of esp32 hanging, even if RSSI became better the esp module can't reconnect again until you restart the esp 32 module. I didn't know if it was the esp 32 or the access point I had that caused the signal to go in that range. to resolve the issue I add in void loop()

void loop(){
  homeSpan.poll();
if( WiFi.isConnected() && WiFi.RSSI()<-90)
     {
        WiFi.disconnect();              
      }
}

**Comment by prolixmusic** [comment]
Interesting.   I should have really good signal - the distance between the AP and the ESP is around 3m (10ft) which is close.  IMO the cadence of the ping responses is very concerning.  When I see a repetitive sawtooth of pings, it makes me believe that my sample rate is close to an even multiple of the loop rate.  Moreover, the fact that the longest pings are approaching 200ms (outliers excluded) It also makes me believe that the poll() loop is taking around 200ms to execute. Idk if this is the case truly the case, but if it is, then I suspect it is starving the Ethernet Stack of processor time.   When I get a chance I may look a little more deeply at the homespan code to see what’s happening.  I may also add a long-press feature to my wall button, to force a hw reset if things go south. That way I won’t have to take the ladder out to reset the esp when it stops responding. Sent from my iPhoneOn Mar 18, 2024, at 7:36 AM, Nezaemmy ***@***.***> wrote:﻿
Hi, I had this issue in the past, When the WiFi RSSI went in the range between -90 and -94 the WiFi module of esp32 hanging, even if RSSI became better the esp module can't reconnect again until you restart the esp 32 module. I didn't know if it was the esp 32 or the access point I had that caused the signal to go in that range. to resolve the issue I add in void loop()
void loop(){
homeSpan.poll();
if( WiFi.isConnected() && WiFi.RSSI()<-90)
{
WiFi.disconnect();
}
}

**Comment by frankonski** [comment]
I just  ran into something similar today. My esp32-on-breadboard dev setup has two "power hardware configuration". 
1. Getting 12v from alarm panel system > 5v regulator (with heatsink) & capacitor for stability > 5v PIN of the esp32 dev board. No USB connection to dev board. I have seen this setup stable, up and running for 2+ weeks. 
2. Getting power via USB to esp32 dev board (and I disconnect that 12v power source)

When I run via USB, it is normally to get access to serial output. 
This morning, my HomeSpan accessories were showing as "Not Responding", and I was in that 2nd / USB powered setup. I could VNC into the laptop which as the dev esp32 board connected, with PIO permanently in _serial monitor_ mode.

There were no errors in the console. Hit Enter, HomeSpan displayed its standard DB info, and the accessories in the Home app became responsive.

The last log in the serial console was ~24 hours ago, which isn't normal, as I know there should have been plenty of more serial lines printed (by a different library).

To me it looks like the board must have gone in low power mode of some sort.

.. another strange (possibly unrelated) thing, is that I use a callback when NTP time is updated, which happens every 4 hours. I print the time received from the callback, compare with the time of my system, and adjust the time of my system accordingly. But this last time entry logged, the NTP time **received** from the NTP server was 1.5 hour **off**... That is not normal either...

**Comment by frankonski** [comment]
.. and now, it lost WiFi, status led keeps flashing once per second, and serial does not reponds to any commands. I checked the power state of the device and USB hub, it is in D0 (full power).

Not sure what's going on...

**Comment by prolixmusic** [comment]
I haven’t had a chance to look at the library in more depth yet.  I did however, check the round trip ping times of a board that I designed and wrote the firmware for. My hardware (running a pic18f97j60 8 bit processor) consistently returns pings with 4ms round trip times.  This is how it ought to be.  The results with homespan are wildly different, hence I get the feeling that the homespan library is taking is starving the Ethernet stack.  My ESP homespan project has been running now for a couple of weeks, after previously suffering from three of four failures back to back. That said, I still think the pings are very telling… when I get the chance I may see if I can split the  library into smaller sub tasks , interspersed with stack calls.  I am pretty sure that allowing the stack to be more responsive will make the library more robust.  Sent from my iPhone. On Mar 28, 2024, at 11:17 AM, frankonski ***@***.***> wrote:﻿
.. and now, it lost WiFi, status led keeps flashing once per second, and serial does not reponds to any commands. I checked the power state of the device and USB hub, it is in D0 (full power).
Not sure what's going on...

**Comment by HomeSpan** [comment]
No need to guess - you can run any of the WiFi examples that Espressif provides.  I just tried the *SimpleWiFiServer* example and the results of a ping yield the same pattern as HomeSpan:

```
64 bytes from 192.168.1.11: icmp_seq=323 ttl=255 time=199.605 ms
64 bytes from 192.168.1.11: icmp_seq=324 ttl=255 time=232.263 ms
64 bytes from 192.168.1.11: icmp_seq=325 ttl=255 time=182.654 ms
64 bytes from 192.168.1.11: icmp_seq=326 ttl=255 time=296.353 ms
64 bytes from 192.168.1.11: icmp_seq=327 ttl=255 time=78.652 ms
64 bytes from 192.168.1.11: icmp_seq=328 ttl=255 time=50.929 ms
64 bytes from 192.168.1.11: icmp_seq=329 ttl=255 time=265.951 ms
64 bytes from 192.168.1.11: icmp_seq=330 ttl=255 time=339.872 ms
64 bytes from 192.168.1.11: icmp_seq=331 ttl=255 time=269.596 ms
64 bytes from 192.168.1.11: icmp_seq=332 ttl=255 time=143.144 ms
```

As an additional check, I deleted all the server code from the *SimpleWiFiServer* example to create a sketch that does nothing but establishes a WiFi connection:

```

#include <WiFi.h>

const char* ssid     = "yourSSID";
const char* password = "yourPassword";

void setup()
{

    delay(10);


    WiFi.begin(ssid, password);

    while (WiFi.status() != WL_CONNECTED) {
        delay(500);
    }

}

void loop(){

}
```

Ping results are the same.  Whatever the reason for this pattern, it's unrelated to the HomeSpan library.

**Comment by prolixmusic** [comment]
This is both interesting concerning... Glad to here it's not homespan causing this, but regardless I suspect this may well have some impact on performance and the inexplicable failures some of us seem to be plagued with. I wonder how ESP code compiled outside the Arduino environment performs? I will email Espressif to what they have to say about round trip ping performance graphs I posted previously.

**Comment by HomeSpan** [comment]
That's a good idea.  You may want to start with the folks developing the Arduino-ESP32 library since they seem to be separate from those that develop the core IDF.

**Comment by prolixmusic** [comment]
Of course mine went unresponsive today. LolSent from my iPhoneOn Mar 29, 2024, at 11:38 AM, HomeSpan ***@***.***> wrote:﻿
That's a good idea.  You may want to start with the folks developing the Arduino-ESP32 library since they seem to be separate from those that develop the core IDF.

**Comment by ronny-antoon** [comment]
I have an Idea for the WiFi problems, I think its all about Signal Power and hardware issues.
i have a Mesh Network, and the esp32 connecting to the Wrong Access Point always. i don't know why. so i tried to force the esp32 to connect to the best Access point with the best signal, by the driver codes didn't achieved it., so i made a snap code to force it by MAC Address of the Access Point. 

`
#include <Arduino.h>
#include "WiFi.h"

void setup()
{
Serial.begin(115200);

    // Set WiFi to station mode and disconnect from an AP if it was previously connected.
    delay(100);

    WiFi.disconnect(true, true);
    WiFi.mode(wifi_mode_t::WIFI_MODE_STA);
    WiFi.scanNetworks(false, false, false, 300, 0, "WIFI SSID", NULL);
    
    WiFi.begin(WiFi.SSID(0), "WiFi Password", 0, WiFi.BSSID(0), true);

    Serial.print("Connecting ");
    int i = 0;

    while ((WiFi.status() != WL_CONNECTED) && (i < 15 * 2))
    {
        delay(500);
        Serial.print(".");
        i++;
    }
    Serial.println("!");
    }
`

in this code the esp32 forced to connect to the best Access Point Signal Power

**Comment by prolixmusic** [comment]
Sorry, I have been really busy at work and haven’t had a chance to dig deeper into this.  That said l have a class that I wrote to debounce inputs.  It’s called once per loop and increments or decrements a count based on the status of the input… my guesstimation based on the response time to key presses and the count threshold in my decounce routine, is that that the loop is being serviced somewhere around 10,000 times a second or 100us intervals. Whereas, based on WiFi pings, it seems like the tcpip stack is being serviced at a cadence of less than 50 times a second, which in my opinion is problematic.I don’t have anything with erratic pings like that in my home… someone else did a WiFi only test of the ESP and got similar results. One wonders what priority is given to the tcpip stack task in RTOS. I would like to see consistent sub-6ms pings which would mean servicing tcpip stack tasks much more frequently.  I build a lot of embedded systems using small microprocessors - usually with far less capability than the ESP. Nearly everything I build is Ethernet based.  My boards are consistently sub 3ms ping hardwired and, on a recent test on my home network, sub-5ms when bridged via wifi to the network.  Long story short, IMO the WiFi needs to be more agile than it is, which I think means increasing its task priority in the RTOS.Sent from my iPhoneOn Apr 13, 2024, at 5:56 AM, Ronny M. Antoon ***@***.***> wrote:﻿
I have an Idea for the WiFi problems, I think its all about Signal Power and hardware issues.
i have a Mesh Network, and the esp32 connecting to the Wrong Access Point always. i don't know why. so i tried to force the esp32 to connect to the best Access point with the best signal, by the driver codes didn't achieved it., so i made a snap code to force it by MAC Address of the Access Point.
`
#include <Arduino.h>
#include "WiFi.h"
void setup()
{
Serial.begin(115200);
// Set WiFi to station mode and disconnect from an AP if it was previously connected.
delay(100);

WiFi.disconnect(true, true);
WiFi.mode(wifi_mode_t::WIFI_MODE_STA);
WiFi.scanNetworks(false, false, false, 300, 0, "WIFI SSID", NULL);

WiFi.begin(WiFi.SSID(0), "WiFi Password", 0, WiFi.BSSID(0), true);

Serial.print("Connecting ");
int i = 0;

while ((WiFi.status() != WL_CONNECTED) && (i < 15 * 2))
{
    delay(500);
    Serial.print(".");
    i++;
}
Serial.println("!");
}

`
in this code the esp32 forced to connect to the best Access Point Signal Power

**Comment by frankonski** [comment]
Home app is now also currently in this "No response" state. However:
- Wifi still works
- weblogs (built-in, not my TLS web server version) working fine
- Serial is responding to CLI commands
- Printing json database shows no error in the json format
- No memory error

**Comment by ronny-antoon** [comment]
> Home app is now also currently in this "No response" state. However:
> 
> * Wifi still works
> * weblogs (built-in, not my TLS web server version) working fine
> * Serial is responding to CLI commands
> * Printing json database shows no error in the json format
> * No memory error


Still the same issue, that's really annoying and i had tried every thing i know.
@HomeSpan Can you help me by making a HomeSpan Restart Function.
I have my program running together with the HomeSpan, and i want to make a function that when i click my button it will make a restart for the homespan alone without my program restart.
i want the restart to include restarting  -  MDNS - hapServer - WIFI

Can you help me do that?

**Comment by prolixmusic** [comment]
Same here.  In my case it crashed Twice in 24Hours after working for weeks. I did notice yesterday it took 10 or so seconds to respond to the garage door request to close.  Then when I got home 3 hours later it was non-responsive in the app and had to be reset again. Sent from my iPhoneOn Apr 21, 2024, at 3:12 AM, Ronny M. Antoon ***@***.***> wrote:﻿

Home app is now also currently in this "No response" state. However:

Wifi still works
weblogs (built-in, not my TLS web server version) working fine
Serial is responding to CLI commands
Printing json database shows no error in the json format
No memory error


Still the same issue, that's really annoying and i had tried every thing i know.
@HomeSpan Can you help me by making a HomeSpan Restart Function.
I have my program running together with the HomeSpan, and i want to make a function that when i click my button it will make a restart for the homespan alone without my program restart.
i want the restart to include restarting  -  MDNS - hapServer - WIFI
Can you help me do that?

**Comment by frankonski** [comment]
It hasn’t occurred since I made my previous post. What I need to check if it happens again is to which controller it is connected to (older AppleTV or new one) and try to remember which of these two AppleTV I used last… because ultimately, Home app. or Eve now connects to the controller, not to HomeSpan…

**Comment by HomeSpan** [comment]
I'll double-check, but I think since iOS17 requires a Home Hub, there will always be at least one active socket connection between HomeKit via the Home Hub) and a HomeSpan device (I need to see if this applies even if there are no Characteristics with notification permission).

Assuming that's the case, I can add some (optional) logic to HomeSpan that checks to ensure there is always at least one active socket connection when WiFi is up, and if it finds there are NO active connections, it will close the WiFi connection and re-connect in a way that re-initialized MDNS.

Though not elegant, this may be a viable workaround given that the source of the problem is unknown, but we know things get (temporarily) resolved if the device reboots.

Given that this issue effects only a very small number of HomeSpan users, and it is intermittent (or rare, as @frankonski noted above) I continue to believe the primary cause is an incompatibility between the way MDNS messages are handled by the ESP32 and the way certain routers work under certain circumstances (ranging from WiFi signal, to drop-outs, to pass-through of MDNS messages, etc.).   Regardless of the cause, hopefully the proposed workaround will address the issue.

**Comment by ronny-antoon** [comment]
@HomeSpan Thank you a lot. you had made a good community with a good open source code. Good Job

**Comment by frankonski** [comment]
It is now happening again, live when I am working on it.
HomeSpan stops responding to **serial commands** as well, /status built-in URL is not responding. The board is responding to pings, and I also know that the main loop is working as the other library I use keeps on logging events in the serial console.

Going in Home app, I can see that it is connected to my _new_ AppleTV, which narrows down the possibility of an issue when my old AppleTV get sets as the controller.

**Comment by frankonski** [comment]
I found other `while(1)` preceded by a `LOG0` rather than `Serial.print()`. I will be updating my PR #823 soon with these extra logging. The issue may hide there.

**Comment by dsbeach** [comment]
>Given that this issue effects only a very small number of HomeSpan users, and it is intermittent (or rare, as @frankonski noted above) I continue to believe the primary cause is an incompatibility between the way MDNS messages are handled by the ESP32 and the way certain routers work under certain circumstances (ranging from WiFi signal, to drop-outs, to pass-through of MDNS messages, etc.). Regardless of the cause, hopefully the proposed workaround will address the issue.

Intermittent, but perhaps not that rare. I have several homespan devices that randomly exhibit this behavior. It ‘feels’ like it is associated with WiFi outages, but I’ve been unable to identify a predictable behavior.

David

**Comment by HomeSpan** [comment]
As an initial test, let's see what happens if we force a disconnect and reconnect of WiFi when the device goes into No Response mode. You can do this by adding a custom CLI command follows:

```
new SpanUserCommand('D', " - disconnect WiFi", [](const char *buf){WiFi.disconnect();})
```

Then, when the device goes into No Response mode, type `@D` into the CLI.

**Comment by eilidan** [comment]
I am a student from China and I want to make a temperature and humidity meter. I replaced the simulated temperature changes with code that reads the actual sensor values in example program 12. However, after uploading the program, the device shows no response in the home app. Even the original program of example 12 is unresponsive after being uploaded. Can you tell me where the problem is?Thanks!!!

**Comment by eilidan** [comment]
> I am a student from China and I want to make a temperature and humidity meter. I replaced the simulated temperature changes with code that reads the actual sensor values in example program 12. However, after uploading the program, the device shows no response in the home app. Even the original program of example 12 is unresponsive after being uploaded. Can you tell me where the problem is?Thanks!!!

Sorry for the inconvenience, but I added the Apple TV to my home, but the Apple TV and ESP32 are not connected to the same WIFI, and the iPhone and ESP32 are connected to the same WIFI, which causes the ESP32 to be able to pair with the phone but not receive data. Because the home hub is not on the LAN.

**Comment by HomeSpan** [comment]
Sorry, HomeKit device must be on the same WiFi network as your Home Hub in order to operate correctly.

**Comment by Antti** [comment]
I believe this is still an issue and it was not fixed. I’m able to pair the device, but after that it becomes unresponsive.

**Comment by frankonski** [comment]
Do you use weblogs AND define an NTP server?
or.. does you own app spawns threads that ends up using `WEBLOG` macro?

See issue #899 which explains one scenario I found that can get HomeSpan devices to become unresponsive until reboot.

**Comment by frankonski** [comment]
My bad, didn't read carefully. Answer was provided by HomeSpan maintainer:
> Sorry, HomeKit device must be on the same WiFi network as your Home Hub in order to operate correctly.

Your AppleTV and ESP32 must be connected to the same WiFi network, and your AppleTV must be connected to the internet with an iCloud account for all the magic to happen when you are away on your phone (on a different wifi, or on cellular data).

**Comment by rackommended** [comment]
hi, i am facing the same problem. I use the 13-TargetStates example code without any changes. the problem is that after a random amount of time, sometimes after a few seconds after a reboot, sometimes after a few minutes, the device goes into the "no response" state in the apple home app, and it stays that way until I press a reset.
however, I have a home assistant server running on raspberry pi, and if I add homekit to it as a device, this problem does not occur there. if I define a homekit bridge in HA, to which I add the homekit-homespan device, and add this bridge to the apple home application, the device can still be used without problems in the home app.
unfortunately, I don't have in-depth knowledge, because this is my first attempt with homespan, I just wrote them down as interesting, in case it helps someone in further investigation

**Comment by HomeSpan** [comment]
Does your RPi connect through WiFi to your home network or through an Ethernet cable?  Also, can you confirm you are running a HomeKit Hub, such as an Apple TV Gen 4 or a HomePod?

**Comment by rackommended** [comment]
the rpi is connected via ethernet and I use a homepod mini as a homehub

**Comment by frankonski** [comment]
Few esp32 boards supports 5 GHz.. if your wifi router allows it, set “wifi client connection preference” so to force 2.4 GHz wifi band for that client, set a preferred access point if you have mesh routers, and disable mesh features for it. Connection is very stable this way if your board does not move around the house.
